File: labs/Lab4/scripts/follow_gap.py
#!/usr/bin/env python
from __future__ import print_function
import sys
import math
import numpy as np

# ROS Imports
import rospy
from sensor_msgs.msg import LaserScan
from ackermann_msgs.msg import AckermannDriveStamped

class reactive_follow_gap:
    def __init__(self):
        # Topics & Subscriptions, Publishers
        lidarscan_topic = '/scan'
        drive_topic = '/drive'

        self.lidar_sub = rospy.Subscriber(lidarscan_topic, LaserScan, self.lidar_callback)
        self.drive_pub = rospy.Publisher(drive_topic, AckermannDriveStamped, queue_size=10)
        self.bubble_radius = 1  # safety bubble radius in meters

    def preprocess_lidar(self, ranges):
        # Rejecting high values (e.g., > 3)
        ranges[ranges > 10] = 0
	#ranges[ranges < 0 ] = 0
        return ranges

    def find_max_gap(self, free_space_ranges):
	    """Return the start index and end index of the max gap in free_space_ranges."""
	    # Find indices where obstacles (zero or less values) end or start
	    changes = np.diff(np.concatenate(([0], free_space_ranges > 0, [0])))
	    gap_starts = np.where(changes == 1)[0]
	    gap_ends = np.where(changes == -1)[0] - 1

	    # Calculate lengths of each gap
	    if len(gap_starts) == 0 or len(gap_ends) == 0:
		return 0, 0  # No gaps found

	    gap_lengths = gap_ends - gap_starts + 1
	    max_gap_idx = np.argmax(gap_lengths)  # Find the index of the maximum gap length

	    # Return the start and end indices of the maximum gap
	    return gap_starts[max_gap_idx], gap_ends[max_gap_idx]

    def find_best_point(self, start_i, end_i, ranges):
	"""Return index of best point in range."""
	# Naive approach: Find the furthest point within the max gap
	max_range_idx = np.argmax(ranges[start_i:end_i+1]) + start_i
	return max_range_idx

    def lidar_callback(self, data):
	"""Process each LiDAR scan as per the Follow Gap algorithm & publish an AckermannDriveStamped Message."""
	ranges = np.array(data.ranges)
	proc_ranges = self.preprocess_lidar(ranges)

	# Find the index of the closest point in the processed ranges
	bubble_idx = np.argmin(proc_ranges[proc_ranges>0])
	theta = self.bubble_radius/np.mean(proc_ranges)
	# Calculate the number of points that fit within the bubble_radius
	angle_increment = data.angle_increment
	num_points_in_bubble = int(theta / angle_increment)

	# Set all points within the bubble radius to zero
	start_idx = max(0, bubble_idx - num_points_in_bubble)
	end_idx = min(len(proc_ranges), bubble_idx + num_points_in_bubble)
	proc_ranges[start_idx:end_idx + 1] = 0

	# Find the maximum length gap
	start_idx, end_idx = self.find_max_gap(proc_ranges)

	# Find the best point in the gap
	best_point_idx = self.find_best_point(start_idx, end_idx, proc_ranges)

	# Calculate the angle to the best point
	angle = data.angle_min + best_point_idx * angle_increment

	# Publish Drive message
	drive_msg = AckermannDriveStamped()
	drive_msg.header.stamp = rospy.Time.now()
	drive_msg.drive.steering_angle = angle
	drive_msg.drive.speed = 0.8  # Set a constant speed or adjust based on context
	self.drive_pub.publish(drive_msg)


def main(args):
    rospy.init_node("FollowGap_node", anonymous=True)
    rfgs = reactive_follow_gap()
    rospy.sleep(0.1)
    rospy.spin()

if __name__ == '__main__':
    main(sys.argv)

File: labs/Lab4/scripts/follow_gap.cpp
#include <ros/ros.h>
#include <sensor_msgs/LaserScan.h>
#include <ackermann_msgs/AckermannDriveStamped.h>
#include <vector>
#include <algorithm>
#include <numeric>

class ReactiveFollowGap {
public:
    ReactiveFollowGap() {
        ros::NodeHandle nh;
        lidar_sub = nh.subscribe("/scan", 10, &ReactiveFollowGap::lidarCallback, this);
        drive_pub = nh.advertise<ackermann_msgs::AckermannDriveStamped>("/drive", 10);
        bubble_radius = 0.5; // meters
    }

private:
    ros::Subscriber lidar_sub;
    ros::Publisher drive_pub;
    double bubble_radius;

    void lidarCallback(const sensor_msgs::LaserScan::ConstPtr& scan) {
        std::vector<float> proc_ranges = preprocessLidar(scan->ranges);

        // Finding the minimum range to calculate the bubble index
        auto bubble_idx = std::distance(proc_ranges.begin(), std::min_element(proc_ranges.begin(), proc_ranges.end()));
        double nominal_distance = std::accumulate(proc_ranges.begin(), proc_ranges.end(), 0.0) / proc_ranges.size();
        double theta = bubble_radius / nominal_distance;
        int num_points_in_bubble = static_cast<int>(theta / scan->angle_increment);

        // Zero out the points within the bubble radius
        int start_idx = std::max(0, static_cast<int>(bubble_idx - num_points_in_bubble));
	int end_idx = std::min(static_cast<int>(proc_ranges.size()), static_cast<int>(bubble_idx + num_points_in_bubble));
        std::fill(proc_ranges.begin() + start_idx, proc_ranges.begin() + end_idx, 0.0);

        // Finding the max gap
        int max_start, max_end;
        findMaxGap(proc_ranges, max_start, max_end);

        // Finding the best point in the max gap
        int best_point_idx = findBestPoint(proc_ranges, max_start, max_end);

        // Calculate the angle to the best point
        double angle = scan->angle_min + best_point_idx * scan->angle_increment;

        // Publish the drive message
        ackermann_msgs::AckermannDriveStamped drive_msg;
        drive_msg.header.stamp = ros::Time::now();
        drive_msg.drive.steering_angle = angle;
        drive_msg.drive.speed = 1.0; // Set a constant speed or adjust based on context
        drive_pub.publish(drive_msg);
    }

    std::vector<float> preprocessLidar(const std::vector<float>& ranges) {
        std::vector<float> proc_ranges(ranges.size(), 0);
        // Simple moving average with window size of 5
        for (size_t i = 2; i < ranges.size() - 2; ++i) {
            proc_ranges[i] = (ranges[i-2] + ranges[i-1] + ranges[i] + ranges[i+1] + ranges[i+2]) / 5.0;
        }
        // Thresholding values greater than 3 meters
        for (auto& range : proc_ranges) {
            if (range > 3) range = 0;
        }
        return proc_ranges;
    }

    void findMaxGap(const std::vector<float>& ranges, int& max_start, int& max_end) {
        int max_gap = 0, current_start = -1;
        max_start = 0; max_end = 0;
        for (size_t i = 0; i < ranges.size(); ++i) {
            if (ranges[i] > 0 && current_start < 0) current_start = i; // Start of a new gap
            else if (ranges[i] == 0 && current_start >= 0) { // End of the current gap
                int current_gap = i - current_start;
                if (current_gap > max_gap) {
                    max_gap = current_gap;
                    max_start = current_start;
                    max_end = i - 1;
                }
                current_start = -1;
            }
        }
        // Check if the last gap reaches the end of the range array
        if (current_start >= 0) {
            int current_gap = ranges.size() - current_start;
            if (current_gap > max_gap) {
                max_gap = current_gap;
                max_start = current_start;
                max_end = ranges.size() - 1;
            }
        }
    }

    int findBestPoint(const std::vector<float>& ranges, int start_idx, int end_idx) {
        return std::distance(ranges.begin(), std::max_element(ranges.begin() + start_idx, ranges.begin() + end_idx + 1));
    }
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "follow_gap_node");
    ReactiveFollowGap reactive_follow_gap;
    ros::spin();
    return 0;
}

